src/Controller/ValetToken/ValetTokenController.spec.ts

```diff
@@ -0,0 +1,54 @@
+import 'reflect-metadata'
+import { Request } from 'express'
+import { ContainerConfigLoader } from '../../Bootstrap/Container'
+
+import { ValetTokenController } from './ValetTokenController'
+import { Container } from 'inversify'
+import TYPES from '../../Bootstrap/Types'
+import { CreateValetToken } from '../../Domain/UseCase/CreateValetToken/CreateValetToken'
+import { insufficientPermissionsDto, sufficientPermissionsDto } from '../../Domain/UseCase/CreateValetToken/test/data'
+
+describe('ValetKeyController', () => {
+  let container: Container
+
+  const createController = () => new ValetTokenController(
+    container.get<CreateValetToken>(TYPES.CreateValetToken)
```

Comment:
all dependencies of a Subject Under Test should be mocked

Reply:
Not in an integration test which I would consider this to be. ;P

I always try to avoid mocking with the verbose test framework utilities if I can help it.

I think a unit-testing ideal is a pure function. Once we attach an implicit state of an object (class) and DI on top in combination with spies, mocks, etc. things tend to get ugly, verbose, and hard to maintain very quickly.

And tests take much longer to write. At least for me.

That said, I can refactor to use more Jest, but does look like a bit of work compared to what is there now.

So unless you say it's ok based on the above, I will jestify.

Reply:
In general we put integration/e2e (name it as you will) tests into SNJS - that's the place where we glue everything up and put it to the final tests if it holds water. In the services that compose our ecosystem let's stick to unit tests - what do you think?

Reply:
I would make a distinction between E2E and integration and prefer that the tests related to x live close to x, but if the current strategy is somehow advantageous it might be ok as well. So far though I am not persuaded. :P

Perhaps a middle ground for now might be to keep E2E in SNJS, but integration and unit close. I have no feeling yet how the current system works in practice, so I can't say if things could be better organized, but in the future I would definitely like to look at SNJS to see if they could.

Meantime I will attempt to jestify these tests and see how it goes.

Reply:
What is the difference between `integration` and `e2e` from your perspective?

I think if we want to put `e2e/integration` into server side projects we would be doubling the work we have to do in SNJS. Since SNJS would have to implement E2E tests checking a given feature/functionality and executing given workflows on the server side project. So now you'd have to maintain a given "workflow test" both in SNJS and on server side - whereas on server side it doesn't give much context without the client application being in picture of the test. 

Also E2E/Integration on server side would have to exist at a separate scope aside - being able to execute it aside of unit testing.

Reply:
I would largely align with this perspective: https://testing.googleblog.com/2015/04/just-say-no-to-more-end-to-end-tests.html
A proper E2E test should simulate real user interaction as closely as possible: input goes in and output is checked at the highest-level interface. This may involve simulations, using something like [Selenium](https://www.selenium.dev/). For pure backend I would tend to classify load testing with [Gatling](https://gatling.io/) or [JMeter](https://jmeter.apache.org/) as E2E.

An integration test just integrates two or more units (for some definition of a unit) together and checks if the work in combination. Does not (need to) test a complete flow. So everything inbetween unit and E2E qualifies here.

I haven't yet taken a good look at what is happening in SNJS tests, so I can't really say much here. I am certainly not for unnecessary doubling of work. By nature though, E2E tests will naturally cover some integration tests and those will naturally cover some unit tests.

To me it seems that it's best if by default tests live close to the things they test, but it surely may be more sensible that they don't.

Anyway feels like a high-level discussion, maybe best taken elsewhere or continued at the next opportunity. For now let's not sweat it. I'd like do what has been working for you guys and after some time we'll see if I have any practical ideas for improvement.

Reply:
We can stay with the integration tests you are proposing here - just make sure to separate them so they don't look exactly the same as unit tests (same file name with `*.spec.ts`) and also so that we can run them by for example `yarn test:integration` and add it to the Github Actions worfklows. 

In this particular case this integration tests does not test anything that a unit test couldn't I think. 

If you've got the `UseCase` unit tested and the `Controller` unit tested (with mocked responses from the Use Case) - not sure what value an integration test adds to this particular controller. In general controllers should be skinny IMHO - parse the request into params, execute a use case, form the response and send it.

Reply:
You are right: it isn't worth it at all in this case.

So opted for removing `container.get`s in favor of stubs and test subjects with stubbed dependencies (no Jest mocks though) to keep it simple and avoid introducing unnecessary workflow complexity.

The entire reason why I went with those `container.get`s in the first place was to save myself some test-writing time. ;)